### ISSUE TITLE
Update all columns with where condition

### ISSUE THREAD

**Issue by pslamka** [linq2db/linq2db#1234]
Hi,

I would like to implement row versioning. I only do full row updates, however I cannot use

db.Update(entity);

because I need to check the row version number. I could use 

db.GetTable<EntityType>().Where(x => x.Version == version).Set(..).Set(..).Set(..).Set(..).Update() 

however that is not very elegant and needs to be changed everytime schema changes.

I was looking into implementation of the full row update which already uses where condition on the row identity, so it should be possible to allow for other update filtering. The project is too complicated for me to create a pull request on this issue.

Is there some other solution to have update + where with generic full row update? Any workarounds?
Could support for this be implemented?

Thank you guys very much!

**Comment by sdanyliv** [comment]
@pslamka, you do not need to change our code, but you can write your own extension method which generates chain of `Set` calls and `Update` at the end. It’s alittle bit tricky but possible.
Sample usage can be found here https://github.com/linq2db/linq2db/issues/1172#issuecomment-394308169. Key method is `SetValues`.

From first sign i see that you will need MappingSchema passed to such method to retrieve information about read only properties (identity, pk, etc.)

**Comment by pslamka** [comment]
@sdanyliv Wow! That was some fast answer. I created a dictionary and used the code as it is and it worked great! I tried to do something similar before posting here, where I tried to fabricate the lambda required by the Set method, but failed because the fabricated lambda did not have any type information for the compiler to infer from.  Now I see I can dynamically invoke the proper version of the Set method while providing it with proper types and invoking it with the fabricated lambda! That is mindwaving :)

I still have a few questions if you don't mind answering them:
1. I used Expression.Property before in my code while trying fabricating property access lambda, whereas you used Expression.MakeMemberAccess - what is the difference?
2. To get the property value for the dictionary, I used reflection (PropertyInfo.GetValue()) - is it possible and more efficient to compile the property access lambda and invoke it to get the value?
3. I am puzzled by the GetColumnExpression method - can you compute the property lambda from the column name? How is that possible?

Thank you very much!

**Comment by sdanyliv** [comment]
@pslamka, to save time for both, i have written sample for you in addition to previous sample. Sorry not tested.
```cs
public static IUpdatable<T> SetValues<T>(this IUpdatable<T> source, T obj, MappingSchema schema = null)
{
   schema = schema ?? MappingSchema.Default;
   var descriptor = schema.GetEntityDescriptor(typeof(T));

   var toUpdate = descriptor.Columns.Where(c => !c.SkipOnUpdate);
   var param = Expression.Parameter(typeof(T));

   object current = source;
   foreach (var c in toUpdate)
   {
      if (c.MemberAccessor.Getter != null)
      {
         var value = c.MemberAccessor.Getter(obj);
         var memberAccess = Expression.MakeMemberAccess(param, c.MemberInfo);
         var lambda = Expression.Lambda(memberAccess, param);
         var method = SetMethodInfo.MakeGenericMethod(typeof(T), memberAccess.Type);
         current = method.Invoke(null, new[] {current, lambda, value});
      }
   }

   return (IUpdatable<T>) current;
}
```
We have plans to create some `linq2db.PowerTools` library. But can not find time.

**Comment by pslamka** [comment]
@sdanyliv Thanks! Works like charm ;)

I have added overload for IQueryable but that's all I had to do. 

```csharp
public static IUpdatable<T> SetValues<T>(this IQueryable<T> source, T obj, MappingSchema schema = null)
{
    return source.AsUpdatable().SetValues(obj, schema);
}
```

I have noticed there is a method on column descriptor class called GetValue which I thought could be used, but it seems to also include the conversion step, therefore I'd suggest renaming it to GetConvertedValue if that's worth anything.

Thanks for your time and help!
